db/revision.go

```diff
@@ -22,13 +23,35 @@ import (
 type Body map[string]interface{}
 
 const (
-	BodyDeleted = "_deleted"
-	BodyRev     = "_rev"
-	BodyId      = "_id"
+	BodyDeleted   = "_deleted"
+	BodyRev       = "_rev"
+	BodyId        = "_id"
+	BodyRevisions = "_revisions"
+)
+
+// A revisions property found within a Body.  Expected to be of the form:
+//   Revisions["start"]: int64, starting generation number
+//   Revisions["ids"]: []string, list of digests
+// Used as map[string]interface{} instead of Revisions struct because it's unmarshalled
+// along with Body, and we don't need the overhead of allocating a new object
+type Revisions map[string]interface{}
+
+const (
+	RevisionsStart = "start"
+	RevisionsIds   = "ids"
 )
 
 func (b *Body) Unmarshal(data []byte) error {
-	if err := json.Unmarshal(data, &b); err != nil {
+
+	if len(data) == 0 {
+		b = &Body{}
+		return nil
+	}
+
+	// Use decoder for unmarshalling to preserve large numbers
+	decoder := json.NewDecoder(bytes.NewReader(data))
+	decoder.UseNumber()
+	if err := decoder.Decode(&b); err != nil {
```

Comment:
`b` is already a pointer, taking the address again produces incorrect values when data = `[]byte("null")`

```go
if err := decoder.Decode(b); err != nil {
```

This test reproduces the issue:

```go
func TestBody_Unmarshal(t *testing.T) {
	var b Body

	err := b.Unmarshal([]byte(""))
	assertNoError(t, err, "expected no error when unmarshalling no bytes")
	assert.DeepEquals(t, b, Body{})

	err = b.Unmarshal([]byte("null"))
	assertNoError(t, err, "expected no error when unmarshalling null")
	assert.DeepEquals(t, b, Body(nil))

	err = b.Unmarshal([]byte("{}"))
	assertNoError(t, err, "expected no error when unmarshalling {}")
	assert.DeepEquals(t, b, Body{})

	err = b.Unmarshal([]byte(`{"test":true}`))
	assertNoError(t, err, "expected no error when unmarshalling example body")
	assert.DeepEquals(t, b, Body{"test": true})
}
```

Reply:
Good catch - thanks.